api/kroxylicious-filter-api/src/main/java/io/kroxylicious/proxy/filter/FilterInvokers.java

```diff
@@ -29,36 +32,62 @@ private FilterInvokers() {
      * @param filter the Filter to create an invoker for
      * @return the invoker
      */
-    public static FilterInvoker from(KrpcFilter filter) {
+    static List<FilterAndInvoker> from(KrpcFilter filter) {
+        List<FilterAndInvoker> filterInvokers = invokersForFilter(filter);
         // all invokers are wrapped in safe invoker so that clients can safely call onRequest/onResponse
         // even if the invoker isn't interested in that message.
-        return new SafeInvoker(invokerForFilter(filter));
+        return wrapAllInSafeInvoker(filterInvokers).toList();
     }
 
-    private static FilterInvoker invokerForFilter(KrpcFilter filter) {
+    private static List<FilterAndInvoker> invokersForFilter(KrpcFilter filter) {
+        boolean isCompositeFilter = filter instanceof CompositeFilter;
         boolean isResponseFilter = filter instanceof ResponseFilter;
         boolean isRequestFilter = filter instanceof RequestFilter;
         boolean isAnySpecificFilterInterface = SpecificFilterArrayInvoker.implementsAnySpecificFilterInterface(filter);
-        if (isAnySpecificFilterInterface && (isRequestFilter || isResponseFilter)) {
-            throw unsupportedFilterInstance(filter, "Cannot mix specific message filter interfaces and [RequestFilter|ResponseFilter] interfaces");
-        }
-        if (!isRequestFilter && !isResponseFilter && !isAnySpecificFilterInterface) {
-            throw unsupportedFilterInstance(filter, "KrpcFilter must implement ResponseFilter, RequestFilter or any combination of specific message Filter interfaces");
+        validateFilter(filter, isResponseFilter, isRequestFilter, isAnySpecificFilterInterface, isCompositeFilter);
+        if (isCompositeFilter) {
```

Comment:
The code looks okay, but got me wondering - is there is a way we could turn these into a compile time issue, rather than a runtime ones.  That lead me to https://stackoverflow.com/a/50534086.   It's not a technique I've seen both - still mulling what I think!

Reply:
Thats certainly an interesting technique. A bit strange to have in our public api however might need some finessing 😁

Reply:
Maybe we could revisit whether they aren't allowed to be implemented together, you could instead have an ordering rule like:

```
1. CompositeFilters are handled first
2. then Request/Response filters
3. then Specific-message filters
```

So If my class looks like:

```
A implement CompositeFilter, RequestFilter, ApiVersionRequestFilter {
    getFilters() {
       return List.of(new FetchResponseFilter)
    }
}
```

We would add to the filter chain three filter+invokers `[FetchResponseFilter, RequestFilter, ApiVersionRequestFilter]`

If they need some explicit ordering you could control it with a `CompositeFilter` at the top.

Feels like mutual exclusion is easier to understand though, less subtle issues with filter ordering.

Reply:
that's true, we could allow `A implement CompositeFilter, RequestFilter, ApiVersionRequestFilter` but:

1. From a user's perspective, I wouldn't intuitively know what order Kroxylicious would apply the different filters.  I'd need to read of the docs.
2. From an implementation point of view, I think mutual exclusion leaves us more leeway to refactor the implementation.  I worry about us boxing ourselves into a corner.

Reply:
:+1: will stick with mutual exclusion